### PR TITLE
Change CLI option behavior for OTA-R and OTA-P to revert back to the success case after using the CLI supplied values

### LINKED ISSUE #15656: [ota-provider-app] Add counters for various QueryImageResponse and ApplyImageResponse commands
#### Problem
There is no way right now to simulate real world scenarios where Provider replies Busy, gets the image and then Available or AwaitNextAction and then Proceed

 #### Proposed Solution
Add counters for various QueryImageResponse and ApplyImageResponse commands to simulate sending Busy/NotAvailable for a specified number of times and then sending Available or sending AwaitNextAction for a specified number of times and then sending Proceed.

### PR DESCRIPTION
**Pull request by isiu-apple**
#### Problem
Currently the values supplied in CLI options for OTA-P and OTA-R remain the same for the duration of its runtime. Therefore real world scenarios where Provider replies Busy, and then on subsequent tries UpdateAvailable is not possible.

Fixes: https://github.com/project-chip/connectedhomeip/issues/15656

#### Change overview

Change the behavior of the following CLI supplied values to use the supplied value only once, and then revert back to the success value on subsequent request/responses.

Provider:
```
-q/--queryImageStatus <updateAvailable | busy | updateNotAvailable> 
-a/--applyUpdateAction <proceed | awaitNextAction | discontinue>
-t/--delayedQueryActionTimeSec <time>
-p/--delayedApplyActionTimeSec <time>
-u/--userConsentState <granted | denied | deferred>
```

Requestor:
```
-u/--userConsentState <granted | denied | deferred>
```

#### Testing
Tested various combinations of the above CLI options and ensured that subsequent QueryImage commands succeed and default to the success path.  Also ensured that the delayedQueryActionTimeSec and delayedApplyActionTimeSec values supplied is applied to the first QueryImageResponse and ApplyUpdateResponses only respectively and reverts back to 0 on subsequent QueryImageResponse and ApplyUpdateResponses respectively.

### CODE REVIEW

File: examples/ota-provider-app/linux/README.md [hunk 1]
```diff
@@ -13,19 +13,19 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -y/--ignoreApplyUpdate <num_times_to_ignore>                          | The number of times to ignore the ApplyUpdate Request and not send a response                                                                                                                                                                                                                                                                                                                                                                                                                                               |
+| -a/--applyUpdateAction <proceed \| awaitNextAction \| discontinue>    | Value for the Action field in the ApplyUpdateResponse for the next ApplyUpdateResponse. <br> Value will revert back to proceed on subsequent responses.                                                                                                                                                                                                                                                                                                                                                                     |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:23]:
```suggestion
| -a/--applyUpdateAction <proceed \| awaitNextAction \| discontinue>    | Value for the Action field in the ApplyUpdateResponse. <br> Value will revert back to proceed on subsequent responses.                                                                                                                                                                                                                                                                                                                                                                     |
```

File: examples/ota-provider-app/linux/README.md [hunk 2]
```diff
@@ -13,19 +13,19 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -t/--delayedActionTimeSec <time>                                      | Value in seconds for the DelayedActionTime field in the QueryImageResponse and ApplyUpdateResponse. <br> Value will revert back to 0 seconds on subsequent QueryImage Responses and ApplyUpdate Responses.                                                                                                                                                                                                                                                                                                                    |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:24]: Should we make this 0 for subsequent? Wondering what we couldn't test if this was always set as the same value?

File: examples/ota-provider-app/linux/README.md [hunk 3]
```diff
@@ -13,19 +13,19 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -u/--userConsentState <granted \| denied \| deferred>                 | The following is applied to the next QueryImageResponse and will revert back to the behavior of granted on subsequent responses. <br> Current user consent state which results in various values for Status field in QueryImageResponse <br> Note that -q/--queryImageStatus overrides this option <li> granted: Status field in QueryImageResponse is set to updateAvailable <li> denied: Status field in QueryImageResponse is set to updateNotAvailable <li> deferred: Status field in QueryImageResponse is set to busy |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:25]:
```suggestion
| -u/--userConsentState <granted \| denied \| deferred>                 | The following is applied to the first QueryImageResponse and will revert back to the behavior of granted on subsequent responses. <br> Current user consent state which results in various values for Status field in QueryImageResponse <br> Note that -q/--queryImageStatus overrides this option <li> granted: Status field in QueryImageResponse is set to updateAvailable <li> denied: Status field in QueryImageResponse is set to updateNotAvailable <li> deferred: Status field in QueryImageResponse is set to busy |
```

File: examples/ota-provider-app/linux/main.cpp [hunk 4]
```diff
@@ -299,17 +299,22 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
                              "  -o/--otaImageList <file>\n"
                              "        Path to a file containing a list of OTA images\n"
                              "  -q/--queryImageStatus <updateAvailable | busy | updateNotAvailable>\n"
-                             "        Value for the Status field in the QueryImageResponse\n"
+                             "        Value for the Status field in the QueryImageResponse for the next QueryImageResponse.\n"
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/main.cpp:302]: Please apply same comments in the README here

File: examples/ota-requestor-app/linux/README.md [hunk 5]
```diff
@@ -22,6 +22,7 @@ following command line options are available for the OTA Requestor application.
 | -p/--periodicQueryTimeout <Time in seconds> | Periodic timeout for querying providers in the default OTA provider list. If none or zero is supplied the timeout is set to every 24 hours. |
 | -c/--requestorCanConsent                    | If supplied, the RequestorCanConsent field of the QueryImage command is set to true. Otherwise, the value is determined by the driver.      |
 | -f/--otaDownloadPath <file path>            | If supplied, the OTA image is downloaded to the given fully-qualified file-path. Otherwise, the value defaults to /tmp/test.bin.            |
+| -u/--userConsentState <granted              | denied                                                                                                                                      | deferred> | The following is applied to the next QueryImage and will revert back to the behavior of granted on subsequent QueryImage commands. <li> granted: Authorize OTA requestor to download an OTA image. <li> denied: Forbid OTA requestor to download an OTA image. <li> deferred: Defer obtaining user consent. |
```
Comment (carol-apple) [inline examples/ota-requestor-app/linux/README.md:25]: I think saying revert back is confusing as the concept of revert doesn't make sense. It's not like they got to specify one behavior before this and it reverts back. Maybe better to just mention default values:
```suggestion
| -u/--userConsentState <granted              | denied                                                                                                                                      | deferred> | The user consent state for the first QueryImage command. For all subsequent commands, the value of granted will be used. <li> granted: Authorize OTA requestor to download an OTA image. <li> denied: Forbid OTA requestor to download an OTA image. <li> deferred: Defer obtaining user consent. |
```

File: examples/ota-requestor-app/linux/main.cpp [hunk 6]
```diff
@@ -93,6 +93,8 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
                              "        If supplied, the OTA image is downloaded to the given fully-qualified file-path.\n"
                              "        Otherwise, the value defaults to /tmp/test.bin.\n "
                              "  -u/--userConsentState <granted | denied | deferred>\n"
+                             "        The following is applied to the next QueryImage and will revert back to\n"
```
Comment (carol-apple) [inline examples/ota-requestor-app/linux/main.cpp:96]: Apply whatever is the fix in the README here

File: examples/ota-provider-app/linux/README.md (hunk 2 above)
Comment (isiu-apple) [reply examples/ota-provider-app/linux/README.md:24]: I asked tester before making this change.  He preferred to use the delayed action time just once to test that functionality, and then revert back to 0 to save time for subsequent tests.

File: examples/ota-provider-app/linux/README.md [hunk 7]
```diff
@@ -13,19 +13,19 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -a/--applyUpdateAction <proceed \| awaitNextAction \| discontinue>    | Value for the Action field in the first ApplyUpdateResponse. <br> For all subsequent responses, the value of proceed will be used.                                                                                                                                                                                                                                                                                                                                                                                         |
+| -t/--delayedActionTimeSec <time>                                      | Value in seconds for the DelayedActionTime field in the QueryImageResponse and ApplyUpdateResponse. <br> Value will revert back to 0 seconds on subsequent QueryImage Responses and ApplyUpdate Responses.                                                                                                                                                                                                                                                                                                                 |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:24]: Hmm, actually this is not always true? You don't reset `mDelayedActionTimeSec` until `HandleApplyUpdateRequest` which means the value specified here will keep applying for QueryImageResponse until a path that leads to `HandleApplyUpdateRequest` being called.

For example, if provider was launched with `--queryImageStatus busy --delayedActionTimeSec 300`. Then you'd get:
Requestor->Provider: QueryImage command
Provider->Requestor: QueryImageResponse with Status=Busy, DelayedActionTime=300
Requestor->Provider: QueryImage command
Provider->Requestor: QueryImageResponse with Status=Available, DelayedActionTime=300
Requestor->Provider: ApplyUpdateRequest command
Provider->Requestor: ApplyUpdateResponse with Action=Proceed, DelayedActionTime=300

I don't believe this is the desired behavior? Wonder if we should just have two separate command line args for each respective delayed time?

Comment (isiu-apple) [reply examples/ota-provider-app/linux/README.md:24]: Sure, I'll split this into 2 CLI args then.

File: .github/.wordlist.txt [hunk 8]
```diff
@@ -326,7 +326,8 @@ DefaultSuccess
 definedValue
 DehumidificationControl
 DelayedActionTime
-delayedActionTimeSec
+delayedApplyQueryActionTimeSec
```
Comment (carol-apple) [inline .github/.wordlist.txt:329]: I think this is supposed to be `delayedApplyActionTimeSec`?

File: examples/ota-provider-app/linux/README.md [hunk 9]
```diff
@@ -13,19 +13,20 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -a/--applyUpdateAction <proceed \| awaitNextAction \| discontinue>    | Value for the Action field in the first ApplyUpdateResponse. <br> For all subsequent responses, the value of proceed will be used.                                                                                                                                                                                                                                                                                                                                                                                         |
+| -t/--delayedQueryActionTimeSec <time>                                 | Value in seconds for the DelayedActionTime field in the QueryImageResponse. <br> Value will revert back to 0 seconds on subsequent QueryImage Responses.                                                                                                                                                                                                                                                                                                                                                                   |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:24]:
```suggestion
| -t/--delayedQueryActionTimeSec <time>                                 | Value in seconds for the DelayedActionTime field in the first QueryImageResponse. <br> Value will revert back to 0 seconds on subsequent QueryImage Responses.                                                                                                                                                                                                                                                                                                                                                                   |
```

File: examples/ota-provider-app/linux/README.md [hunk 10]
```diff
@@ -13,19 +13,20 @@ scripts/examples/gn_build_example.sh examples/ota-provider-app/linux out/debug c
...
+| -p/--delayedApplyActionTimeSec <time>                                 | Value in seconds for the DelayedActionTime field in the ApplyUpdateResponse. <br> Value will revert back to 0 seconds on subsequent ApplyUpdate Responses.                                                                                                                                                                                                                                                                                                                                                                 |
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/README.md:25]:
```suggestion
| -p/--delayedApplyActionTimeSec <time>                                 | Value in seconds for the DelayedActionTime field in the first ApplyUpdateResponse. <br> Value will revert back to 0 seconds on subsequent ApplyUpdate Responses.                                                                                                                                                                                                                                                                                                                                                                 |
```

File: examples/ota-provider-app/linux/main.cpp [hunk 11]
```diff
@@ -299,17 +305,24 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
...
                              "        The number of times to ignore the ApplyUpdate Request and not send a response.\n"
                              "  -a/--applyUpdateAction <proceed | awaitNextAction | discontinue>\n"
-                             "        Value for the Action field in the ApplyUpdateResponse\n"
-                             "  -t/--delayedActionTimeSec <time>\n"
-                             "        Value in seconds for the DelayedActionTime field in the QueryImageResponse\n"
-                             "        and ApplyUpdateResponse\n"
+                             "        Value for the Action field in the first ApplyUpdateResponse.\n"
+                             "        For all subsequent responses, the value of proceed will be used.\n"
+                             "  -t/--delayedQueryActionTimeSec <time>\n"
+                             "        Value in seconds for the DelayedActionTime field in the QueryImageResponse.\n"
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/main.cpp:318]:
```suggestion
                             "        Value in seconds for the DelayedActionTime field in the first QueryImageResponse.\n"
```

File: examples/ota-provider-app/linux/main.cpp [hunk 12]
```diff
@@ -299,17 +305,24 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
...
+                             "        Value will revert back to 0 seconds on subsequent QueryImageResponses.\n"
+                             "  -p/--delayedApplyActionTimeSec <time>\n"
+                             "        Value in seconds for the DelayedActionTime field in the ApplyUpdateResponse.\n"
```
Comment (carol-apple) [inline examples/ota-provider-app/linux/main.cpp:321]:
```suggestion
                             "        Value in seconds for the DelayedActionTime field in the first ApplyUpdateResponse.\n"
```

**Review by carol-apple — APPROVED** [review]

**Review by selissia — APPROVED** [review]

**Review by harsha-rajendran — APPROVED** [review]

**Pull request merged**